### PR TITLE
ISSUE-378 Contact auto-complete

### PR DESCRIPTION
**Pull request by vttranlina**

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]

**Comment by chibenwa** [comment]
Do you have a capture of the gatling run (html) to share already here?

**Comment by chibenwa** [comment]
Even logs of the gatling run.

**Comment by Arsnael** [comment]
I doubt you can run that on preprod, as it extends a normal simulation, inserting new users from scratch (and we are using a ldap read only on preprod so...), not reusing the existing ones

**Comment by vttranlina** [comment]
> we are using a ldap read only on preprod so...), not reusing the existing ones

How can we pass it? Using csv feeder?

**Comment by Arsnael** [comment]
Extending the normal simulation is fine (sorry a bit hard for me today) but yes you need to use the csv feeder, like in jmap draft `PlatformValidationSimulation` for example :)

**Comment by vttranlina** [comment]
> Do you have a capture of the gatling run (html) to share already here?

On my local env, tmail memory server. 

screencapture-file-home-tungtv-workplace-2-JAMES-james-gatling-results-tmailcontactautocompletesimulation-20220413093531477-req-retrieveaccount-26cf4-html-2022-04-13-16_46_19





screencapture-file-home-tungtv-workplace-2-JAMES-james-gatling-results-tmailcontactautocompletesimulation-20220413093531477-index-html-2022-04-13-16_45_52

**Comment by chibenwa** [comment]
Ok. Can you upload the ZIP file of the latest simulation?

The dataset is very small thus we will not have such a wow effect in real life (we fit in RAM, and JVM likely do magics here!)

**Comment by chibenwa** [comment]
Was it an in-memory james server? If so could you try quickly a distributed james server please? (even local...)

**Comment by vttranlina** [comment]
## Benchmark result 
- Env: local
- Distributed James server
- Elasticsearch `user_contact` stats
<details>
  <summary>Click to expand </summary>
  
  ```json
{
  "_shards": {
    "total": 10,
    "successful": 5,
    "failed": 0
  },
  "_all": {
    "primaries": {
      "docs": {
        "count": 516,
        "deleted": 13
      },
      "store": {
        "size_in_bytes": 603656,
        "reserved_in_bytes": 0
      },
      "indexing": {
        "index_total": 530,
        "index_time_in_millis": 506,
        "index_current": 0,
        "index_failed": 0,
        "delete_total": 0,
        "delete_time_in_millis": 0,
        "delete_current": 0,
        "noop_update_total": 0,
        "is_throttled": false,
        "throttle_time_in_millis": 0
      },
      "get": {
        "total": 0,
        "time_in_millis": 0,
        "exists_total": 0,
        "exists_time_in_millis": 0,
        "missing_total": 0,
        "missing_time_in_millis": 0,
        "current": 0
      },
      "search": {
        "open_contexts": 0,
        "query_total": 16281850,
        "query_time_in_millis": 882203,
        "query_current": 0,
        "fetch_total": 88279,
        "fetch_time_in_millis": 7242,
        "fetch_current": 0,
        "scroll_total": 0,
        "scroll_time_in_millis": 0,
        "scroll_current": 0,
        "suggest_total": 0,
        "suggest_time_in_millis": 0,
        "suggest_current": 0
      },
      "merges": {
        "current": 0,
        "current_docs": 0,
        "current_size_in_bytes": 0,
        "total": 22,
        "total_time_in_millis": 706,
        "total_docs": 824,
        "total_size_in_bytes": 2057709,
        "total_stopped_time_in_millis": 0,
        "total_throttled_time_in_millis": 0,
        "total_auto_throttle_in_bytes": 104857600
      },
      "refresh": {
        "total": 260,
        "total_time_in_millis": 2965,
        "external_total": 250,
        "external_total_time_in_millis": 3343,
        "listeners": 0
      },
      "flush": {
        "total": 10,
        "periodic": 0,
        "total_time_in_millis": 34
      },
      "warmer": {
        "current": 0,
        "total": 245,
        "total_time_in_millis": 24
      },
      "query_cache": {
        "memory_size_in_bytes": 0,
        "total_count": 0,
        "hit_count": 0,
        "miss_count": 0,
        "cache_size": 0,
        "cache_count": 0,
        "evictions": 0
      },
      "fielddata": {
        "memory_size_in_bytes": 0,
        "evictions": 0
      },
      "completion": {
        "size_in_bytes": 0
      },
      "segments": {
        "count": 27,
        "memory_in_bytes": 58636,
        "terms_memory_in_bytes": 38016,
        "stored_fields_memory_in_bytes": 13176,
        "term_vectors_memory_in_bytes": 0,
        "norms_memory_in_bytes": 5184,
        "points_memory_in_bytes": 0,
        "doc_values_memory_in_bytes": 2260,
        "index_writer_memory_in_bytes": 0,
        "version_map_memory_in_bytes": 0,
        "fixed_bit_set_memory_in_bytes": 0,
        "max_unsafe_auto_id_timestamp": -1,
        "file_sizes": {}
      },
      "translog": {
        "operations": 0,
        "size_in_bytes": 275,
        "uncommitted_operations": 0,
        "uncommitted_size_in_bytes": 275,
        "earliest_last_modified_age": 691602
      },
      "request_cache": {
        "memory_size_in_bytes": 0,
        "evictions": 0,
        "hit_count": 0,
        "miss_count": 0
      },
      "recovery": {
        "current_as_source": 0,
        "current_as_target": 0,
        "throttle_time_in_millis": 0
      }
    },
    "total": {
      "docs": {
        "count": 516,
        "deleted": 13
      },
      "store": {
        "size_in_bytes": 603656,
        "reserved_in_bytes": 0
      },
      "indexing": {
        "index_total": 530,
        "index_time_in_millis": 506,
        "index_current": 0,
        "index_failed": 0,
        "delete_total": 0,
        "delete_time_in_millis": 0,
        "delete_current": 0,
        "noop_update_total": 0,
        "is_throttled": false,
        "throttle_time_in_millis": 0
      },
      "get": {
        "total": 0,
        "time_in_millis": 0,
        "exists_total": 0,
        "exists_time_in_millis": 0,
        "missing_total": 0,
        "missing_time_in_millis": 0,
        "current": 0
      },
      "search": {
        "open_contexts": 0,
        "query_total": 16281850,
        "query_time_in_millis": 882203,
        "query_current": 0,
        "fetch_total": 88279,
        "fetch_time_in_millis": 7242,
        "fetch_current": 0,
        "scroll_total": 0,
        "scroll_time_in_millis": 0,
        "scroll_current": 0,
        "suggest_total": 0,
        "suggest_time_in_millis": 0,
        "suggest_current": 0
      },
      "merges": {
        "current": 0,
        "current_docs": 0,
        "current_size_in_bytes": 0,
        "total": 22,
        "total_time_in_millis": 706,
        "total_docs": 824,
        "total_size_in_bytes": 2057709,
        "total_stopped_time_in_millis": 0,
        "total_throttled_time_in_millis": 0,
        "total_auto_throttle_in_bytes": 104857600
      },
      "refresh": {
        "total": 260,
        "total_time_in_millis": 2965,
        "external_total": 250,
        "external_total_time_in_millis": 3343,
        "listeners": 0
      },
      "flush": {
        "total": 10,
        "periodic": 0,
        "total_time_in_millis": 34
      },
      "warmer": {
        "current": 0,
        "total": 245,
        "total_time_in_millis": 24
      },
      "query_cache": {
        "memory_size_in_bytes": 0,
        "total_count": 0,
        "hit_count": 0,
        "miss_count": 0,
        "cache_size": 0,
        "cache_count": 0,
        "evictions": 0
      },
      "fielddata": {
        "memory_size_in_bytes": 0,
        "evictions": 0
      },
      "completion": {
        "size_in_bytes": 0
      },
      "segments": {
        "count": 27,
        "memory_in_bytes": 58636,
        "terms_memory_in_bytes": 38016,
        "stored_fields_memory_in_bytes": 13176,
        "term_vectors_memory_in_bytes": 0,
        "norms_memory_in_bytes": 5184,
        "points_memory_in_bytes": 0,
        "doc_values_memory_in_bytes": 2260,
        "index_writer_memory_in_bytes": 0,
        "version_map_memory_in_bytes": 0,
        "fixed_bit_set_memory_in_bytes": 0,
        "max_unsafe_auto_id_timestamp": -1,
        "file_sizes": {}
      },
      "translog": {
        "operations": 0,
        "size_in_bytes": 275,
        "uncommitted_operations": 0,
        "uncommitted_size_in_bytes": 275,
        "earliest_last_modified_age": 691602
      },
      "request_cache": {
        "memory_size_in_bytes": 0,
        "evictions": 0,
        "hit_count": 0,
        "miss_count": 0
      },
      "recovery": {
        "current_as_source": 0,
        "current_as_target": 0,
        "throttle_time_in_millis": 0
      }
    }
  },
  "indices": {
    "user_contact": {
      "uuid": "_naJAuOoS7i0KrBuvhgmrA",
      "primaries": {
        "docs": {
          "count": 516,
          "deleted": 13
        },
        "store": {
          "size_in_bytes": 603656,
          "reserved_in_bytes": 0
        },
        "indexing": {
          "index_total": 530,
          "index_time_in_millis": 506,
          "index_current": 0,
          "index_failed": 0,
          "delete_total": 0,
          "delete_time_in_millis": 0,
          "delete_current": 0,
          "noop_update_total": 0,
          "is_throttled": false,
          "throttle_time_in_millis": 0
        },
        "get": {
          "total": 0,
          "time_in_millis": 0,
          "exists_total": 0,
          "exists_time_in_millis": 0,
          "missing_total": 0,
          "missing_time_in_millis": 0,
          "current": 0
        },
        "search": {
          "open_contexts": 0,
          "query_total": 16281850,
          "query_time_in_millis": 882203,
          "query_current": 0,
          "fetch_total": 88279,
          "fetch_time_in_millis": 7242,
          "fetch_current": 0,
          "scroll_total": 0,
          "scroll_time_in_millis": 0,
          "scroll_current": 0,
          "suggest_total": 0,
          "suggest_time_in_millis": 0,
          "suggest_current": 0
        },
        "merges": {
          "current": 0,
          "current_docs": 0,
          "current_size_in_bytes": 0,
          "total": 22,
          "total_time_in_millis": 706,
          "total_docs": 824,
          "total_size_in_bytes": 2057709,
          "total_stopped_time_in_millis": 0,
          "total_throttled_time_in_millis": 0,
          "total_auto_throttle_in_bytes": 104857600
        },
        "refresh": {
          "total": 260,
          "total_time_in_millis": 2965,
          "external_total": 250,
          "external_total_time_in_millis": 3343,
          "listeners": 0
        },
        "flush": {
          "total": 10,
          "periodic": 0,
          "total_time_in_millis": 34
        },
        "warmer": {
          "current": 0,
          "total": 245,
          "total_time_in_millis": 24
        },
        "query_cache": {
          "memory_size_in_bytes": 0,
          "total_count": 0,
          "hit_count": 0,
          "miss_count": 0,
          "cache_size": 0,
          "cache_count": 0,
          "evictions": 0
        },
        "fielddata": {
          "memory_size_in_bytes": 0,
          "evictions": 0
        },
        "completion": {
          "size_in_bytes": 0
        },
        "segments": {
          "count": 27,
          "memory_in_bytes": 58636,
          "terms_memory_in_bytes": 38016,
          "stored_fields_memory_in_bytes": 13176,
          "term_vectors_memory_in_bytes": 0,
          "norms_memory_in_bytes": 5184,
          "points_memory_in_bytes": 0,
          "doc_values_memory_in_bytes": 2260,
          "index_writer_memory_in_bytes": 0,
          "version_map_memory_in_bytes": 0,
          "fixed_bit_set_memory_in_bytes": 0,
          "max_unsafe_auto_id_timestamp": -1,
          "file_sizes": {}
        },
        "translog": {
          "operations": 0,
          "size_in_bytes": 275,
          "uncommitted_operations": 0,
          "uncommitted_size_in_bytes": 275,
          "earliest_last_modified_age": 691602
        },
        "request_cache": {
          "memory_size_in_bytes": 0,
          "evictions": 0,
          "hit_count": 0,
          "miss_count": 0
        },
        "recovery": {
          "current_as_source": 0,
          "current_as_target": 0,
          "throttle_time_in_millis": 0
        }
      },
      "total": {
        "docs": {
          "count": 516,
          "deleted": 13
        },
        "store": {
          "size_in_bytes": 603656,
          "reserved_in_bytes": 0
        },
        "indexing": {
          "index_total": 530,
          "index_time_in_millis": 506,
          "index_current": 0,
          "index_failed": 0,
          "delete_total": 0,
          "delete_time_in_millis": 0,
          "delete_current": 0,
          "noop_update_total": 0,
          "is_throttled": false,
          "throttle_time_in_millis": 0
        },
        "get": {
          "total": 0,
          "time_in_millis": 0,
          "exists_total": 0,
          "exists_time_in_millis": 0,
          "missing_total": 0,
          "missing_time_in_millis": 0,
          "current": 0
        },
        "search": {
          "open_contexts": 0,
          "query_total": 16281850,
          "query_time_in_millis": 882203,
          "query_current": 0,
          "fetch_total": 88279,
          "fetch_time_in_millis": 7242,
          "fetch_current": 0,
          "scroll_total": 0,
          "scroll_time_in_millis": 0,
          "scroll_current": 0,
          "suggest_total": 0,
          "suggest_time_in_millis": 0,
          "suggest_current": 0
        },
        "merges": {
          "current": 0,
          "current_docs": 0,
          "current_size_in_bytes": 0,
          "total": 22,
          "total_time_in_millis": 706,
          "total_docs": 824,
          "total_size_in_bytes": 2057709,
          "total_stopped_time_in_millis": 0,
          "total_throttled_time_in_millis": 0,
          "total_auto_throttle_in_bytes": 104857600
        },
        "refresh": {
          "total": 260,
          "total_time_in_millis": 2965,
          "external_total": 250,
          "external_total_time_in_millis": 3343,
          "listeners": 0
        },
        "flush": {
          "total": 10,
          "periodic": 0,
          "total_time_in_millis": 34
        },
        "warmer": {
          "current": 0,
          "total": 245,
          "total_time_in_millis": 24
        },
        "query_cache": {
          "memory_size_in_bytes": 0,
          "total_count": 0,
          "hit_count": 0,
          "miss_count": 0,
          "cache_size": 0,
          "cache_count": 0,
          "evictions": 0
        },
        "fielddata": {
          "memory_size_in_bytes": 0,
          "evictions": 0
        },
        "completion": {
          "size_in_bytes": 0
        },
        "segments": {
          "count": 27,
          "memory_in_bytes": 58636,
          "terms_memory_in_bytes": 38016,
          "stored_fields_memory_in_bytes": 13176,
          "term_vectors_memory_in_bytes": 0,
          "norms_memory_in_bytes": 5184,
          "points_memory_in_bytes": 0,
          "doc_values_memory_in_bytes": 2260,
          "index_writer_memory_in_bytes": 0,
          "version_map_memory_in_bytes": 0,
          "fixed_bit_set_memory_in_bytes": 0,
          "max_unsafe_auto_id_timestamp": -1,
          "file_sizes": {}
        },
        "translog": {
          "operations": 0,
          "size_in_bytes": 275,
          "uncommitted_operations": 0,
          "uncommitted_size_in_bytes": 275,
          "earliest_last_modified_age": 691602
        },
        "request_cache": {
          "memory_size_in_bytes": 0,
          "evictions": 0,
          "hit_count": 0,
          "miss_count": 0
        },
        "recovery": {
          "current_as_source": 0,
          "current_as_target": 0,
          "throttle_time_in_millis": 0
        }
      }
    }
  }
}
   ```
</details>


- Gatling result: 
[tmailcontactautocompletesimulation-20220414043410205.zip](https://github.com/linagora/james-gatling/files/8486324/tmailcontactautocompletesimulation-20220414043410205.zip)
- Simulation
```scala
  setUp(scenario.generate(
    duration = (10 minutes),
    userFeeder = feederFactory.userFeeder(),
    recipientFeeder = feederFactory.recipientFeeder())
    .inject(rampUsers(UserCount) during (1 minute))
    .protocols(HttpSettings.httpProtocol))
```

**Comment by chibenwa** [comment]
Screenshot from 2022-04-14 14-20-40


Awesome!

File: src/test/scala-2.12/org/apache/james/gatling/simulation/UsersFeederWebAdminFactory.scala [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+import io.gatling.core.Predef._
+import org.apache.james.gatling.control.RecipientFeeder.RecipientFeederBuilder
+import org.apache.james.gatling.control.UserFeeder.UserFeederBuilder
+import org.apache.james.gatling.control.{RecipientFeeder, User, UserCreator, UserFeeder}
+
+import scala.concurrent.ExecutionContext.Implicits.global
+import scala.concurrent.duration.Duration.Inf
+import scala.concurrent.{Await, Future}
+
+class UsersFeederWebAdminFactory(userCount: Int = Configuration.UserCount) {
```
Comment (Arsnael) [inline src/test/scala-2.12/org/apache/james/gatling/simulation/UsersFeederWebAdminFactory.scala:12]: You forgot to refactor the code following this or it's in a following commit?

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**